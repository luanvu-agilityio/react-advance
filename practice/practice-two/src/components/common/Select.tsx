import * as SelectPrimitive from '@radix-ui/react-select'
import styled from 'styled-components'
import { ChevronDown, Check } from 'lucide-react'

export interface SelectOption {
  value: string
  label: string
  disabled?: boolean
}

export interface SelectProps {
  options: SelectOption[]
  value?: string
  onChange?: (value: string) => void
  placeholder?: string
  label?: string
  disabled?: boolean

  required?: boolean
  name?: string
  className?: string
  variant?: 'default' | 'search'
  style?: React.CSSProperties
}

const SelectContainer = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
`

const Label = styled.label`
  font-size: 14px;
  color: #374151;
  margin-bottom: 6px;
  font-weight: 500;
`

const StyledTrigger = styled(SelectPrimitive.Trigger)<{
  error?: boolean
  variant?: string
}>`
  display: inline-flex;
  align-items: center;
  justify-content: space-between;
  border-radius: ${(props) => (props.variant === 'search' ? '12px' : '4px')};
  padding: ${(props) => (props.variant === 'search' ? '0' : '10px 12px')};
  font-size: ${(props) => (props.variant === 'search' ? '15px' : '16px')};
  font-weight: ${(props) => (props.variant === 'search' ? '700' : 'normal')};
  line-height: 1;
  gap: 8px;
  background-color: ${(props) =>
    props.variant === 'search' ? 'transparent' : 'white'};
  color: ${(props) => (props.variant === 'search' ? '#000' : '#1F2937')};
  border: ${(props) => {
    if (props.variant === 'search') return 'none'
    const borderColor = props.error ? '#EF4444' : '#D1D5DB'
    return `1px solid ${borderColor}`
  }};
  height: ${(props) => (props.variant === 'search' ? '100%' : 'auto')};
  width: 100%;
  cursor: pointer;

  &:focus {
    outline: none;
    border-color: ${(props) => (props.error ? '#EF4444' : '#4F46E5')};
    box-shadow: ${(props) => {
      if (props.variant === 'search') return 'none'
      const shadowColor = props.error
        ? 'rgba(239, 68, 68, 0.2)'
        : 'rgba(79, 70, 229, 0.2)'
      return `0 0 0 3px ${shadowColor}`
    }};
  }

  &[data-placeholder] {
    color: #9ca3af;
  }

  &:disabled {
    opacity: 0.5;
    cursor: not-allowed;
  }
`

const StyledContent = styled(SelectPrimitive.Content)`
  z-index: 1000;
  position: relative;
  top: 10px;
  left: -20px;
  overflow: hidden;
  background-color: white;
  border-radius: 4px;
  box-shadow:
    0px 10px 38px -10px rgba(22, 23, 24, 0.35),
    0px 10px 20px -15px rgba(22, 23, 24, 0.2);
  min-width: var(--radix-select-trigger-width);
  max-height: 300px;
`

const StyledViewport = styled(SelectPrimitive.Viewport)`
  padding: 5px;
`

const StyledItem = styled(SelectPrimitive.Item)`
  font-size: 15px;
  line-height: 1;
  color: #1f2937;
  border-radius: 3px;
  display: flex;
  align-items: center;
  padding: 10px 35px 10px 10px;
  position: relative;
  user-select: none;
  cursor: pointer;

  &:focus {
    outline: none;
    background-color: #f3f4f6;
  }

  &[data-disabled] {
    color: #9ca3af;
    pointer-events: none;
  }

  &[data-highlighted] {
    background-color: #f3f4f6;
  }
`

const StyledItemIndicator = styled(SelectPrimitive.ItemIndicator)`
  position: absolute;
  right: 10px;
  width: 25px;
  display: inline-flex;
  align-items: center;
  justify-content: center;
`

const StyledItemText = styled(SelectPrimitive.ItemText)`
  margin-left: 20px;
`

const StyledIcon = styled(ChevronDown)<{ variant?: string }>`
  color: ${(props) => (props.variant === 'search' ? '#6B8F40' : '#6B7280')};
`

const StyledScrollUpButton = styled(SelectPrimitive.ScrollUpButton)`
  display: flex;
  align-items: center;
  justify-content: center;
  height: 25px;
  background-color: white;
  color: #6b7280;
  cursor: pointer;
`

const StyledScrollDownButton = styled(SelectPrimitive.ScrollDownButton)`
  display: flex;
  align-items: center;
  justify-content: center;
  height: 25px;
  background-color: white;
  color: #6b7280;
  cursor: pointer;
`

const Select = ({
  options,
  value,
  onChange,
  label,
  disabled = false,

  required = false,
  name,
  className,
  variant = 'default',
  placeholder = 'Select an option',
  style,
}: SelectProps) => {
  return (
    <SelectContainer className={className} style={style}>
      {label && (
        <Label>
          {label}
          {required && ' *'}
        </Label>
      )}

      <SelectPrimitive.Root
        value={value}
        onValueChange={onChange}
        disabled={disabled}
        name={name}
      >
        <StyledTrigger variant={variant}>
          <SelectPrimitive.Value placeholder={placeholder} />
          <SelectPrimitive.Icon>
            <StyledIcon
              size={16}
              variant={variant}
              style={{ strokeWidth: '4' }}
            />
          </SelectPrimitive.Icon>
        </StyledTrigger>

        <SelectPrimitive.Portal>
          <StyledContent position="popper">
            <StyledScrollUpButton>
              <ChevronDown size={16} style={{ transform: 'rotate(180deg)' }} />
            </StyledScrollUpButton>

            <StyledViewport>
              {options.map((option) => (
                <StyledItem
                  key={option.value}
                  value={option.value}
                  disabled={option.disabled}
                >
                  <StyledItemIndicator>
                    <Check size={16} />
                  </StyledItemIndicator>
                  <StyledItemText>{option.label}</StyledItemText>
                </StyledItem>
              ))}
            </StyledViewport>

            <StyledScrollDownButton>
              <ChevronDown size={16} />
            </StyledScrollDownButton>
          </StyledContent>
        </SelectPrimitive.Portal>
      </SelectPrimitive.Root>
    </SelectContainer>
  )
}

export default Select
