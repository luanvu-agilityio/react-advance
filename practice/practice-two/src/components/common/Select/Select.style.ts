import styled from 'styled-components'
import * as SelectPrimitive from '@radix-ui/react-select'
import { ChevronDown } from 'lucide-react'

export const SelectContainer = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
`

export const Label = styled.label`
  font-size: 14px;
  color: #374151;
  margin-bottom: 6px;
  font-weight: 500;
`

export const StyledTrigger = styled(SelectPrimitive.Trigger)<{
  error?: boolean
  variant?: string
}>`
  display: inline-flex;
  align-items: center;
  border-radius: ${(props) => (props.variant === 'search' ? '12px' : '4px')};
  padding: 0;
  font-size: ${(props) => (props.variant === 'search' ? '15px' : '14px')};
  font-weight: ${(props) => (props.variant === 'search' ? '700' : 'normal')};
  line-height: 1;
  gap: 8px;
  background-color: transparent;
  color: ${(props) => (props.variant === 'search' ? '#000' : '#1F2937')};
  border: none;

  height: ${(props) => (props.variant === 'search' ? '100%' : 'auto')};
  width: 100%;
  cursor: pointer;

  &:focus {
    outline: none;
    border-color: ${(props) => (props.error ? '#EF4444' : '#4F46E5')};
    box-shadow: ${(props) => {
      if (props.variant === 'search') return 'none'
      const shadowColor = props.error
        ? 'rgba(239, 68, 68, 0.2)'
        : 'rgba(79, 70, 229, 0.2)'
      return `0 0 0 3px ${shadowColor}`
    }};
  }

  &[data-placeholder] {
    color: #9ca3af;
  }

  &:disabled {
    opacity: 0.5;
    cursor: not-allowed;
  }
`

export const StyledContent = styled(SelectPrimitive.Content)`
  z-index: 1000;
  position: relative;
  top: 10px;
  left: -20px;
  overflow: hidden;
  background-color: white;
  border-radius: 4px;
  box-shadow:
    0px 10px 38px -10px rgba(22, 23, 24, 0.35),
    0px 10px 20px -15px rgba(22, 23, 24, 0.2);
  min-width: var(--radix-select-trigger-width);
  max-height: 300px;
`

export const StyledViewport = styled(SelectPrimitive.Viewport)`
  padding: 5px;
`

export const StyledItem = styled(SelectPrimitive.Item)`
  font-size: 15px;
  line-height: 1;
  color: #1f2937;
  border-radius: 3px;
  display: flex;
  align-items: center;
  padding: 10px 35px 10px 10px;
  position: relative;
  user-select: none;
  cursor: pointer;

  &:focus {
    outline: none;
    background-color: #f3f4f6;
  }

  &[data-disabled] {
    color: #9ca3af;
    pointer-events: none;
  }

  &[data-highlighted] {
    background-color: #f3f4f6;
  }
`

export const StyledItemIndicator = styled(SelectPrimitive.ItemIndicator)`
  position: absolute;
  right: 10px;
  width: 25px;
  display: inline-flex;
  align-items: center;
  justify-content: center;
`

export const StyledItemText = styled(SelectPrimitive.ItemText)`
  margin-left: 20px;
`

export const StyledIcon = styled(ChevronDown)<{ variant?: string }>`
  color: ${(props) => (props.variant === 'search' ? '#6B8F40' : '#6B7280')};
`

export const StyledScrollUpButton = styled(SelectPrimitive.ScrollUpButton)`
  display: flex;
  align-items: center;
  justify-content: center;
  height: 25px;
  background-color: white;
  color: #6b7280;
  cursor: pointer;
`

export const StyledScrollDownButton = styled(SelectPrimitive.ScrollDownButton)`
  display: flex;
  align-items: center;
  justify-content: center;
  height: 25px;
  background-color: white;
  color: #6b7280;
  cursor: pointer;
`
