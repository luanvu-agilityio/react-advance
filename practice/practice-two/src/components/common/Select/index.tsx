import * as SelectPrimitive from '@radix-ui/react-select'

import { ChevronDown, Check } from 'lucide-react'
import {
  SelectContainer,
  StyledContent,
  StyledIcon,
  StyledItem,
  StyledItemIndicator,
  StyledItemText,
  StyledScrollDownButton,
  StyledScrollUpButton,
  StyledTrigger,
  StyledViewport,
} from './Select.style'
import { Label } from '@radix-ui/react-form'
import type { CSSProperties } from 'react'

export interface SelectOption {
  value: string
  label: string
  disabled?: boolean
}

export interface SelectProps {
  options: SelectOption[]
  value: string
  onChange: (value: string) => void
  placeholder?: string
  label?: string
  disabled?: boolean
  required?: boolean
  name?: string
  className?: string
  variant?: 'default' | 'search' | 'filter'
  style?: CSSProperties
}

const Select = ({
  options,
  value,
  onChange,
  label,
  disabled = false,
  required = false,
  name,
  className,
  variant = 'default',
  placeholder = 'Select an option',
  style,
}: SelectProps) => {
  return (
    <SelectContainer className={className} style={style}>
      {label && (
        <Label>
          {label}
          {required && ' *'}
        </Label>
      )}

      <SelectPrimitive.Root
        value={value}
        onValueChange={onChange}
        disabled={disabled}
        name={name}
      >
        <StyledTrigger variant={variant} className="rt-SelectTrigger">
          <SelectPrimitive.Value placeholder={placeholder} />
          <SelectPrimitive.Icon>
            <StyledIcon
              size={16}
              variant={variant}
              style={{ strokeWidth: '4' }}
            />
          </SelectPrimitive.Icon>
        </StyledTrigger>

        <SelectPrimitive.Portal>
          <StyledContent position="popper" className="rt-SelectContent">
            <StyledScrollUpButton className="rt-SelectScrollUpButton">
              <ChevronDown size={16} style={{ transform: 'rotate(180deg)' }} />
            </StyledScrollUpButton>

            <StyledViewport className="rt-SelectViewport">
              {options.map((option) => (
                <StyledItem
                  key={option.value}
                  value={option.value}
                  disabled={option.disabled}
                  className="rt-SelectItem"
                >
                  <StyledItemIndicator className="rt-SelectItemIndicator">
                    <Check size={16} />
                  </StyledItemIndicator>
                  <StyledItemText className="rt-SelectItemText">
                    {option.label}
                  </StyledItemText>
                </StyledItem>
              ))}
            </StyledViewport>

            <StyledScrollDownButton className="rt-SelectScrollDownButton">
              <ChevronDown size={16} />
            </StyledScrollDownButton>
          </StyledContent>
        </SelectPrimitive.Portal>
      </SelectPrimitive.Root>
    </SelectContainer>
  )
}
export default Select
