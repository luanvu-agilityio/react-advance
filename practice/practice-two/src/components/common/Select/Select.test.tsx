import { render, screen, fireEvent, within } from '@testing-library/react'
import Select from './index'
import type { ReactNode } from 'react'

jest.mock('@radix-ui/react-form', () => ({
  Form: ({ children }: { children: ReactNode }) => (
    <form data-testid="radix-form">{children}</form>
  ),
  Label: ({
    children,
    ...props
  }: {
    children: ReactNode
    [key: string]: unknown
  }) => (
    <label data-testid="radix-label" {...props}>
      {children}
    </label>
  ),
}))

// Mock portal behavior for Radix UI components
jest.mock('@radix-ui/react-select', () => {
  const actual = jest.requireActual('@radix-ui/react-select')

  return {
    ...actual,
    Portal: ({ children }: { children: ReactNode }) => (
      <div data-testid="select-portal">{children}</div>
    ),
  }
})

describe('Select Component', () => {
  const mockOptions = [
    { value: 'option1', label: 'Option 1' },
    { value: 'option2', label: 'Option 2' },
    { value: 'option3', label: 'Option 3', disabled: true },
  ]

  const mockOnChange = jest.fn()

  const defaultProps = {
    options: mockOptions,
    onChange: mockOnChange,
    placeholder: 'Select an option',
    value: '',
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  // SNAPSHOT TESTS FOR RENDERING
  it('renders different states correctly', () => {
    // Default state
    const { container: defaultContainer } = render(<Select {...defaultProps} />)
    expect(defaultContainer).toMatchSnapshot('default')

    // With label
    const { container: labelContainer } = render(
      <Select {...defaultProps} label="Test Label" />
    )
    expect(labelContainer).toMatchSnapshot('with-label')

    // Disabled
    const { container: disabledContainer } = render(
      <Select {...defaultProps} disabled />
    )
    expect(disabledContainer).toMatchSnapshot('disabled')
  })

  // INTERACTIVE BEHAVIOR TESTS
  describe('Interactive behavior', () => {
    it('calls onChange when an option is selected', () => {
      render(<Select {...defaultProps} />)

      // Open the select dropdown
      const trigger = screen.getByRole('combobox')
      fireEvent.click(trigger)

      //
      const portal = screen.getByTestId('select-portal')

      // Find option within the portal
      const option = within(portal).getByText('Option 1')
      fireEvent.click(option)

      expect(mockOnChange).toHaveBeenCalledWith('option1')
    })

    it('renders disabled option as non-interactive', () => {
      render(<Select {...defaultProps} />)

      // Open the select dropdown
      const trigger = screen.getByRole('combobox')
      fireEvent.click(trigger)

      // Find the portal
      const portal = screen.getByTestId('select-portal')

      // Get the disabled option and click it
      const disabledOption = within(portal).getByText('Option 3')
      fireEvent.click(disabledOption)

      // Verify the onChange was not called
      expect(mockOnChange).not.toHaveBeenCalled()
    })

    it('closes when option is selected', () => {
      const { rerender } = render(<Select {...defaultProps} />)

      // Open dropdown
      const trigger = screen.getByRole('combobox')
      fireEvent.click(trigger)

      // Portal should be visible
      expect(screen.getByTestId('select-portal')).toBeInTheDocument()

      // Select option
      const option = within(screen.getByTestId('select-portal')).getByText(
        'Option 2'
      )
      fireEvent.click(option)

      // Rerender with the new value to simulate state update
      rerender(<Select {...defaultProps} value="option2" />)

      // Click elsewhere to ensure dropdown closed
      fireEvent.mouseDown(document.body)

      expect(screen.queryByText('Option 2')).toBeInTheDocument()
    })
  })
})
