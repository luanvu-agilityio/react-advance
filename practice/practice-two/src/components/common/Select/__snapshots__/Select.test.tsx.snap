// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Component renders different states correctly: default 1`] = `
<div>
  <div
    class="sc-Qotzb fDwOxq"
  >
    <button
      aria-autocomplete="none"
      aria-controls="radix-«r0»"
      aria-expanded="false"
      class="sc-dKREkF jkuJWC"
      data-placeholder=""
      data-state="closed"
      dir="ltr"
      role="combobox"
      type="button"
      variant="default"
    >
      <span
        style="pointer-events: none;"
      >
        Select an option
      </span>
      <span
        aria-hidden="true"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-down sc-icnseD hKzWei"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          style="stroke-width: 4;"
          variant="default"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m6 9 6 6 6-6"
          />
        </svg>
      </span>
    </button>
    <div
      data-testid="select-portal"
    />
  </div>
</div>
`;

exports[`Select Component renders different states correctly: disabled 1`] = `
<div>
  <div
    class="sc-Qotzb fDwOxq"
  >
    <button
      aria-autocomplete="none"
      aria-controls="radix-«r8»"
      aria-expanded="false"
      class="sc-dKREkF jkuJWC"
      data-disabled=""
      data-placeholder=""
      data-state="closed"
      dir="ltr"
      disabled=""
      role="combobox"
      type="button"
      variant="default"
    >
      <span
        style="pointer-events: none;"
      >
        Select an option
      </span>
      <span
        aria-hidden="true"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-down sc-icnseD hKzWei"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          style="stroke-width: 4;"
          variant="default"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m6 9 6 6 6-6"
          />
        </svg>
      </span>
    </button>
    <div
      data-testid="select-portal"
    />
  </div>
</div>
`;

exports[`Select Component renders different states correctly: with-label 1`] = `
<div>
  <div
    class="sc-Qotzb fDwOxq"
  >
    <label
      data-testid="radix-label"
    >
      Test Label
    </label>
    <button
      aria-autocomplete="none"
      aria-controls="radix-«r4»"
      aria-expanded="false"
      class="sc-dKREkF jkuJWC"
      data-placeholder=""
      data-state="closed"
      dir="ltr"
      role="combobox"
      type="button"
      variant="default"
    >
      <span
        style="pointer-events: none;"
      >
        Select an option
      </span>
      <span
        aria-hidden="true"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-down sc-icnseD hKzWei"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          style="stroke-width: 4;"
          variant="default"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m6 9 6 6 6-6"
          />
        </svg>
      </span>
    </button>
    <div
      data-testid="select-portal"
    />
  </div>
</div>
`;
